### PR TITLE
 ONNX export in `.train()` mode fix

### LINKED ISSUE #3346: models/export.py setting model.train() may be changed to false in onnx
## 🐛 Bug

I think you may want to have no grid construction in [Detect](https://github.com/ultralytics/yolov5/blob/aad99b63d6ac63278021a66f8a096e5232ed24b0/models/yolo.py#L50) layer  by setting [model.train()](https://github.com/ultralytics/yolov5/blob/aad99b63d6ac63278021a66f8a096e5232ed24b0/models/export.py#L62) when export model to onnx with --train. But in "torch.onnx.export", model.training seems to be reset to mode.training=False by default.  

In fact, the param 'training' in the "torch.onnx.export" can be set to "training=torch.onnx.TrainingMode.TRAINING", so that the model can be set to train for conversion instead of using model.train().  But it doesn't seem to be a good way to be recommended. And I found that this part was correct before version 4.0, but after version 5.0, the "export" was deleted in Detect. 

## To Reproduce (REQUIRED)

Input:
```
python models/export.py --weights yolov5s.pt --img 640 --batch 1 --train
```

Output:
```
TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if self.grid[i].shape[2:4] != x[i].shape[2:4] or self.onnx_dynamic:
```

## Expected behavior
This part should be skipped. https://github.com/ultralytics/yolov5/blob/aad99b63d6ac63278021a66f8a096e5232ed24b0/models/yolo.py#L50-L62

## Environment

 - OS: Ubuntu18.04
 - GPU: 3080
 - PyTorch 1.8.0+cu111
 -  onnx 1.9.0

### PR DESCRIPTION
**Pull request by ChaofWang**
This PR is submitted for this bug #3346

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced ONNX export options for different training modes.

### 📊 Key Changes
- 🔄 Added an option to specify training or evaluation mode during ONNX export.
- 🧱 Enabled or disabled constant folding depending on the selected mode (training or evaluation).

### 🎯 Purpose & Impact
- 🎛 This improvement allows developers to export their models in the proper mode, matching their intended use case (either for further training or for inference).
- 🚀 Users can now optimize their models' ONNX export for training with gradient updates or for faster inference without gradients.
- 🛠 The option to toggle constant folding can help in optimizing the model accordingly, potentially improving performance in inference scenarios.

### CODE REVIEW

**Review by glenn-jocher — APPROVED** [review]

**Comment by glenn-jocher** [comment]
@ChaofWang PR is merged, thank you for your contributions!
